### PR TITLE
ddp forward support custom backend.

### PR DESCRIPTION
**Pull request by huihoaan**
Currently DDP only considers CUDA backend，DDP forward will transfer tensor to CUDA. We want ddp to run on custom backend.

### CODE REVIEW

**Review by ezyang — CHANGES_REQUESTED** [review]
Can we make this just accept a device instead?

File: torch/distributed/utils.py [hunk 1]
```diff
@@ -49,26 +49,26 @@ def _unpack_kwargs(flat_args: Tuple[Any, ...], kwarg_keys: Tuple[str, ...]) -> T
...
+            if device == target_device:
                 return (obj,)
             if not use_side_stream_for_tensor_copies:
-                return (obj.to(target_gpu),)
+                return (obj.to(target_device),)
             else:
                 # Perform CPU -> GPU copies in a background stream. This code is
                 # motivated from similar logic in torch/nn/parallel/_functions.py
-                stream = _get_stream(target_gpu)
+                stream = _get_stream(target_device.index)
```
Comment (ezyang) [inline torch/distributed/utils.py:67]: Does this actually work for your use case?

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Need lintfix

**Comment by huihoaan** [comment]
> Can we make this just accept a device instead?

do you mean, use DistributedDataParallel.device_type and device_ids to structure a torch.device, then to comple with input? 
could you taka second look on my pr? @ezyang

File: torch/distributed/utils.py (hunk 1 above)
Comment (huihoaan) [reply torch/distributed/utils.py:67]: > Does this actually work for your use case?
Currently，in my use case, input is already on my backend.  But how do I use stream is on my plane.

**Comment by huihoaan** [comment]
> Need lintfix

yes, i will fix lint next day. Today is mainly about communication ideas.

**Comment by ezyang** [comment]
Updated PR looks good.

**Comment by huihoaan** [comment]
I have fixed the lint,please merge this pr  @ezyang

**Comment by huihoaan** [comment]
I found my test case failed because of this pr：
https://github.com/pytorch/pytorch/pull/98216

could you take a look? @mrshenli  this is the error message：
File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/distributed/_spmd/api.py", line 387, in <module>
aten.all_reduce.default,
File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/_ops.py", line 713, in __getattr__
raise AttributeError(
     AttributeError: '_OpNamespace' 'aten' object has no attribute 'all_reduce'

**Comment by huihoaan** [comment]
@ezyang this pr run ut failed, error message: 
`RuntimeError: Failed to import torch.distributed._spmd.api: '_OpNamespace' 'aten' object has no attribute 'all_reduce'`
It wasn't caused by my code, I have create a new DDP pr, which just change variable name, also have the same error. test pr:https://github.com/pytorch/pytorch/pull/98588
what can I do now?

**Comment by huihoaan** [comment]
> @pytorchbot merge -r

`merge -r` still failed in onnx_test_ut, colud you type`merge -f`?

**Pull request closed without merging**